crates/net/network/src/transactions/mod.rs

```diff
@@ -1082,84 +1083,193 @@ where
+                    Poll::Ready(Some(event)) => {
+                        this.on_network_event(event);
+
+                        maybe_more_work = true;
+
+                        budget_network_events -= 1;
+                        if budget_network_events == 0 {
+                            break 'network_events
+                        }
+                    }
+                    Poll::Ready(None) => {} // todo: handle stream closed as error
+                    Poll::Pending => break 'network_events,
+                }
+            }
 
-        // drain incoming transaction events
-        while let Poll::Ready(Some(event)) = this.transaction_events.poll_next_unpin(cx) {
-            this.on_network_tx_event(event);
-        }
+            // drain commands
+            let mut budget_commands = 1;
+            'commands: loop {
+                match this.command_rx.poll_next_unpin(cx) {
+                    Poll::Ready(Some(cmd)) => {
+                        this.on_command(cmd);
 
-        this.update_request_metrics();
+                        maybe_more_work = true;
 
-        // drain fetching transaction events
-        while let Poll::Ready(Some(fetch_event)) = this.transaction_fetcher.poll_next_unpin(cx) {
-            match fetch_event {
-                FetchEvent::TransactionsFetched { peer_id, transactions } => {
-                    this.import_transactions(peer_id, transactions, TransactionSource::Response);
+                        budget_commands -= 1;
+                        if budget_commands == 0 {
+                            break 'commands
+                        }
+                    }
+                    Poll::Ready(None) => {} // todo: handle stream closed as error
+                    Poll::Pending => break 'commands,
                 }
-                FetchEvent::FetchError { peer_id, error } => {
-                    trace!(target: "net::tx", ?peer_id, ?error, "requesting transactions from peer failed");
-                    this.on_request_error(peer_id, error);
+            }
+
+            // drain incoming transaction events
+            let mut budget_incoming_tx_events = 1;
+            'incoming_tx_events: loop {
+                match this.transaction_events.poll_next_unpin(cx) {
+                    Poll::Ready(Some(event)) => {
+                        this.on_network_tx_event(event);
+
+                        maybe_more_work = true;
+
+                        budget_incoming_tx_events -= 1;
+                        if budget_incoming_tx_events == 0 {
+                            break 'incoming_tx_events
+                        }
+                    }
+                    Poll::Ready(None) => {} // todo: handle stream closed as error
+                    Poll::Pending => break 'incoming_tx_events,
                 }
             }
-        }
 
-        // try drain buffered transactions
-        this.request_buffered_hashes();
+            this.update_request_metrics();
```

Comment:
do we want to update this metric before and after poll the tx fetcher for events? this is how it is on main

Reply:
@mattsse ? last remaining comment